LFS-961: Cannot specify custom ethnicities

veronikaslc opened the pull request.

Commit:
LFS-961: Cannot specify custom ethnicities

marta- commented on modules/data-entry/src/main/frontend/src/vocabQuery/VocabularyQuery.jsx:
```diff
@@ -240,19 +242,57 @@ function VocabularyQuery(props) {
+        >
+          <Typography
+              component="p"
+              color="textSecondary"
```
The disabled item + textSecondary color turns into a very light gray. Let's get rid of this line to make the text more visible.
```suggestion
```

marta- replied:
Done in 882eeec

marta- commented on modules/data-entry/src/main/frontend/src/vocabQuery/VocabularyQuery.jsx:
```diff
@@ -240,19 +242,57 @@ function VocabularyQuery(props) {
-          {NO_RESULTS_TEXT}
+          <Typography
+              component="p"
+              color="textSecondary"
```
Same suggestion as above.
```suggestion
```

marta- replied:
Done in 882eeec

marta- commented on modules/data-entry/src/main/frontend/src/vocabQuery/queryStyle.jsx:
```diff
@@ -109,6 +109,15 @@ const thesaurusStyle = theme => ({
+  noneOfAboveResultsText: {
+    marginBottom: theme.spacing(-1),
+  },
+  noneOfAboveResults: {
+    borderTop: "2px solid " + grayColor[5]
+  }
```
Replace this styling with a `<Divider />` inserted between the list of suggestions and the "None of the above" option.

```suggestion
```
Also, with the introduction of the `Divider`, the two classes above can now be merged to simplify styling.

marta- replied:
Done in 882eeec

marta- committed:
LFS-961: Cannot specify custom ethnicities
Simplified styling

marta- committed:
LFS-961: Cannot specify custom ethnicities
Simplified code, fixed console warning

marta- commented:
Confusing behavior for **maxAnswers=1**:
* Type and select any non-predefined value (e.g. "Other") and save
* Type something else in the input (e.g. Unknown) and **leave the input without clicking** on the option indicated by "No results, use..." -> the value of the input remains as "Unknown"
* Save and close -> observe how "Unknown" was not saved; the saved value is "Other"

In other words, what you see is not what you get.

I see two options for single select:
* either restore the original value if the user leaves the input without clicking on "No results, use..."
* or allow to save the value entered regardless of the interaction eith the dropdown.

Commit:
LFS-961: Cannot specify custom ethnicities
restore the original value if the user leaves the input without option selection

marta- commented on modules/data-entry/src/main/frontend/src/vocabQuery/VocabularyQuery.jsx:
```diff
@@ -310,6 +310,7 @@ function VocabularyQuery(props) {
       && maxAnswers !== 1
       && clearOnClick
       && setInputValue("");
+    maxAnswers == 1 && initialSelection?.length > 0 && setInputValue(initialSelection[0][0]);
```
initialSelection[0][LABEL_POS]

Commit:
LFS-961: Cannot specify custom ethnicities
restore the original value if the user leaves the input without option selectionBetter code fore readability

sdumitriu commented:
Weird things are happening when using a vocabulary term as free text, e.g.:

- lfs mode
- install HANCESTRO
- new Patient Information form
- far Race type `Swazi` (with an uppercase S), but instead of selecting the matching vocabulary term, select it as a custom term

Then trying to replace the custom term with the vocabulary one is difficult:
- trying to select `Swazi` from the suggestion doesn't work
- trying to select it from the browser doesn't work
- in the vocabulary browser, first deleting the custom term then selecting the vocabulary term will clear the custom term but fail to add the vocabulary term
- finally, once the custom term is removed from the list, the vocabulary term can be selected

Not sure if this is a big priority to fix, a reasonable scenario that users may actually encounter is using a custom term that is later added to the vocabulary, and the user goes back and tries to update the term. @veronikaslc is this easy to fix? @marta- is it important to fix?

sdumitriu commented:
Another problem, with single select inputs:
- change the Race question to have maxAnswers=1
- search for a custom term and select it
- using the radio buttons choose one of the predefined race questions
- bug: the input always reflects the selected term
- selecting the radio button next to the input fortunately redisplays the previously selected custom term

marta- commented:
> @veronikaslc is this easy to fix? @marta- is it important to fix?

It should be investigated. It's not a blocker, but if we can fix it let's

marta- commented:
> Weird things are happening when using a vocabulary term as free text, e.g.:
> 
> * lfs mode
> * install HANCESTRO
> * new Patient Information form
> * far Race type `Swazi` (with an uppercase S), but instead of selecting the matching vocabulary term, select it as a custom term
> 
> Then trying to replace the custom term with the vocabulary one is difficult:
> 
> * trying to select `Swazi` from the suggestion doesn't work
> * trying to select it from the browser doesn't work
> * in the vocabulary browser, first deleting the custom term then selecting the vocabulary term will clear the custom term but fail to add the vocabulary term
> * finally, once the custom term is removed from the list, the vocabulary term can be selected

In the MultipleChoice component there a few places where we check if either the id OR the label of a newly added answer matches something already selected. Could that be part of the cause of this bug? @veronikaslc

veronikaslc commented:
> Another problem, with single select inputs:
> 
> * change the Race question to have maxAnswers=1
> * search for a custom term and select it
> * using the radio buttons choose one of the predefined race questions
> * bug: the input always reflects the selected term
> * selecting the radio button next to the input fortunately redisplays the previously selected custom term

Please specify the expected behaviour

veronikaslc commented:
> > Weird things are happening when using a vocabulary term as free text, e.g.:
> > 
> > * lfs mode
> > * install HANCESTRO
> > * new Patient Information form
> > * far Race type `Swazi` (with an uppercase S), but instead of selecting the matching vocabulary term, select it as a custom term
> > 
> > Then trying to replace the custom term with the vocabulary one is difficult:
> > 
> > * trying to select `Swazi` from the suggestion doesn't work
> > * trying to select it from the browser doesn't work
> > * in the vocabulary browser, first deleting the custom term then selecting the vocabulary term will clear the custom term but fail to add the vocabulary term
> > * finally, once the custom term is removed from the list, the vocabulary term can be selected
> 
> In the MultipleChoice component there a few places where we check if either the id OR the label of a newly added answer matches something already selected. Could that be part of the cause of this bug? @veronikaslc

Yes, its is part of a problem indeed

Commit:
LFS-961: Cannot specify custom ethnicities
restore the original value if the user leaves the input without option selectionFixed same label term manipulations

sashaandjic commented:
> > Another problem, with single select inputs:
> > 
> > * change the Race question to have maxAnswers=1
> > * search for a custom term and select it
> > * using the radio buttons choose one of the predefined race questions
> > * bug: the input always reflects the selected term
> > * selecting the radio button next to the input fortunately redisplays the previously selected custom term
> 
> Please specify the expected behaviour

Once the predefined option is selected keep the free text as is (what was originally typed) and unselected.

sashaandjic commented:
Is it just me, or the dialog buttons are looking funny?
![image](https://user-images.githubusercontent.com/23243244/124701806-2bcd6b80-debd-11eb-9205-f638a1fbf9d1.png)

sashaandjic commented:
Now with the ability to enter free text, we should better handle situation when the vocabulary isn't installed yet.
![image](https://user-images.githubusercontent.com/23243244/124701953-76e77e80-debd-11eb-9b75-94daa7178f41.png)
 
Perhaps, we should default to this one in any case:
![image](https://user-images.githubusercontent.com/23243244/124702089-caf26300-debd-11eb-98af-84ff950a2afc.png)

sashaandjic commented:
Now I can select free text "Swazi" and vocabulary term "Swazi" - "HANCESTRO_0546"
![image](https://user-images.githubusercontent.com/23243244/124702959-633d1780-debf-11eb-84dd-5dd3f2866303.png)

If free text is matching an existing term, can we not offer selection of the free text by not displaying "No results, use..."?
In the sample below, it does not make sense to say "No results, use Lao"
![image](https://user-images.githubusercontent.com/23243244/124703199-d8a8e800-debf-11eb-84d8-d5c4663f04a5.png)

veronikaslc commented:
@marta- Let's summarize all of the expected behaviour from the last 4 comments.

marta- commented:
> Is it just me, or the dialog buttons are looking funny?

They sure look funny in the screenshot, but I wasn't able to replicate the problem.

marta- commented:
@veronikaslc Summary of outstanding fixes:

#### Regression
For vocabulary questions with **maxAnswers=1** with list+input display, whenever one of the default options is selected, the label of that option is copied into the input.
**Expected behavior:** the input should remain empty when clicking on a predefined option

#### Usability issue 1
Currently if any of the vocabularies associated with a vocabulary question is not installed, we display an error message under the input ("Cannot load answer suggestions...") and prevent the user from entering any data. That was ok before implementing the ability to enter free text.
**Expected behavior:** continue to display the error message, but allow the user to enter their own answer by displaying the dropdown anyway with just the one option "No results, use \<input value\>".
Additional improvement, if easily implemented here: if an input is linked to more than one vocabulary, and at least one vocabulary is installed:
* provide suggestions from that vocabulary
* display the error message "Some answer suggestions for this question could not be loaded. Please inform your administrator".
* log a message in the JS console specifying the vocabulary or vocabularies that are missing: "Cannnot load answer suggestions from ORDO".

#### Usability issue 2
When the user's input matches the label of a suggestion **exactly**, displaying the same value twice looks like a bug (even though it actually isn't).
**Expected behavior:** Check for exact matched between the input value and the suggestion labels. If there is an exact match, do not display the "None of the above, use \<input value\>" option, which will appear redundant.

veronikaslc commented:
> provide suggestions from the vocabulary or vocabularies that are installed

That is already implemented

marta- commented:
> > provide suggestions from the vocabulary or vocabularies that are installed
> 
> That is already implemented

Just clarifying that displaying the error message doesn't mean we don't show available suggestions. Sorry for the confusion.

This improvement is now logged separately as https://phenotips.atlassian.net/browse/LFS-1234 .

marta- commented:
> #### Usability issue 1
> Currently if any of the vocabularies associated with a vocabulary question is not installed, we display an error message under the input ("Cannot load answer suggestions...") and prevent the user from entering any data. That was ok before implementing the ability to enter free text.
> **Expected behavior:** continue to display the error message, but allow the user to enter their own answer by displaying the dropdown anyway with just the one option _"No results, use <input value>_".

Done in 5ddc42a .

> #### Usability issue 2
> When the user's input matches the label of a suggestion **exactly**, displaying the same option twice (once as a suggestion from the vocabulary, and once at the bottom of the dropdown as the user's own input marked with "_None of the above..._") looks like a bug, even though it actually isn't.
> **Expected behavior:** Check for exact matches between the input value and the suggestion labels. If there is an exact match, do not display the _"None of the above, use <input value>_" option, which will appear redundant.

Done in 5ddc42a .

marta- committed:
LFS-961: Cannot specify custom ethnicities
* allow user to enter their own text when the vocabulary is not installed
* don't display the user input option in the dropdown if it matches one of the suggestions exactly

veronikaslc commented:
> Just clarifying that displaying the error message doesn't mean we don't show available suggestions. Sorry for the confusion.

I am abit confused here. The way it is working right now - if there are any suggestions loaded from any installed available vocab - we show them without the error. What I mean is that we are already prividing suggestions from the vocabulary or vocabularies that are installed.

marta- commented:
> > Just clarifying that displaying the error message doesn't mean we don't show available suggestions. Sorry for the confusion.
> 
> I am abit confused here. The way it is working right now - if there are any suggestions loaded from any installed available vocab - we show them without the error. What I mean is that we are already prividing suggestions from the vocabulary or vocabularies that are installed.

Rephrased as _"the suggestions from the vocabulary or vocabularies that are installed should continue to be displayed as they are now_", hope this helps.

marta- commented:
> #### Regression
> For vocabulary questions with **maxAnswers=1** with list+input display, whenever one of the default options is selected, the label of that option is copied into the input.
> **Expected behavior:** the input should remain empty when clicking on a predefined option

Should be fixed in f2b9865, although I'd like less duplicate code across `VocabularyQuery`.

marta- committed:
LFS-961: Cannot specify custom ethnicities
Fixed regression where the input is mirroring the label of selected default options when maxAnswers == 1

marta- commented:
> Should be fixed in [f2b9865](https://github.com/ccmbioinfo/lfs/commit/f2b9865767df73867922c4605dcccd6ca1443dd9), although I'd like less duplicate code across `VocabularyQuery`.

Commit f891a8e consists of some refactoring and hopefully does not introduce any regressions, @sashaandjic please test to confirm.

marta- committed:
LFS-961: Cannot specify custom ethnicities
Refactoring to reduce code duplication

sashaandjic commented:
Looks and feels good. All previous complaints have been resolved.
I am not sure why the buttons are looking really weird:
![image](https://user-images.githubusercontent.com/23243244/124871107-a49df780-df91-11eb-8955-10c81fb4614d.png)

Also, a very minor thing, that is **very unlikely to happen**, but may confuse the user.
![image](https://user-images.githubusercontent.com/23243244/124871212-c26b5c80-df91-11eb-834f-da3a67cf1629.png)

marta- commented:
> I am not sure why the buttons are looking really weird:
I can't see any changes on this branch that could affect the look of the buttons. Do they look right on other branches?

> Also, a very minor thing, that is **very unlikely to happen**, but may confuse the user.
> If vocabulary isn't installed yet, we may better handle the messaging. Perhaps, we can get rid of message displayed in red fonts.

Ok, we can get rid of the message if we agree it isn't important to inform the user that the suggestions aren't working properly.

veronikaslc commented:
> > I am not sure why the buttons are looking really weird:
> 
> I can't see any changes on this branch that could affect the look of the buttons. Do they look right on other branches?
> 
> > Also, a very minor thing, that is **very unlikely to happen**, but may confuse the user.
> > If vocabulary isn't installed yet, 
> Ok, we can get rid of the message if we agree it isn't important to inform the user that the suggestions aren't working properly.

User won't see the full error text, it will always be covered with the suggestion of the input.

marta- commented:
> Ok, we can get rid of the message if we agree it isn't important to inform the user that the suggestions aren't working properly.

Alternatively we can display the error message in the dropdown where the suggestions were supposed to be, Wdyt @sashaandjic @veronikaslc ?
![image](https://user-images.githubusercontent.com/651980/124936361-75f54080-dfd4-11eb-9d74-34f2f5477c2a.png)

marta- commented:
> Alternatively we can display the error message in the dropdown where the suggestions were supposed to be, Wdyt @sashaandjic @veronikaslc ?

Done in de02a22, we can revert it if there is disagreement.

marta- committed:
LFS-961: Cannot specify custom ethnicities
Ensure the error message informing the user that suggestions can't be loaded is actually visible

sashaandjic reviewed (approved):
Tested- all good!
![image](https://user-images.githubusercontent.com/23243244/124950154-2a489400-dfe0-11eb-9d17-94f2d9120632.png)

veronikaslc commented:
Seems close to be merged. I'm starting LFS-1234 on top of this one then.

sdumitriu reviewed (approved):
I have a few minor issues that will be reported separately.
